File: internal/validate/validate.go
// Package validate performs per-driver syntax/character checks on a query before
// it runs. It pairs the officially-maintained command/method sets (Redis,
// MongoDB) with a universal balanced-delimiter scan. (A real SQL grammar parser
// was tried but rejected: the available parser false-positives on valid
// MySQL/Doris queries, so only the never-false-positive checks remain.) Issues
// are advisory: the UI surfaces them as a non-blocking warning and may colour the
// offending span.
package validate

import (
	"strings"

	"tdb/internal/config"
	"tdb/internal/suggest"
)

type Severity int

const (
	SevWarning Severity = iota
	SevError
)

// Issue is a single problem located in the query text by byte offset/length.
type Issue struct {
	Offset   int
	Length   int
	Severity Severity
	Message  string
}

// Validate returns the issues found in text for the given driver. An empty slice
// means no problems were detected.
func Validate(driver config.Driver, text string) []Issue {
	if strings.TrimSpace(text) == "" {
		return nil
	}
	// A balanced-delimiter scan applies to every driver and never false-positives
	// on valid input.
	issues := unbalancedDelimiters(text)
	switch driver {
	case config.DriverRedis:
		issues = append(issues, validateRedis(text)...)
	case config.DriverMongo:
		issues = append(issues, validateMongo(text)...)
	}
	return issues
}

func validateRedis(text string) []Issue {
	commands := suggest.RedisCommandSet()
	var issues []Issue
	offset := 0
	for _, line := range strings.SplitAfter(text, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" && !strings.HasPrefix(trimmed, "#") {
			fields := strings.Fields(trimmed)
			cmd := strings.ToUpper(fields[0])
			if !commands[cmd] {
				start := offset + strings.Index(line, fields[0])
				issues = append(issues, Issue{
					Offset:   start,
					Length:   len(fields[0]),
					Severity: SevWarning,
					Message:  "redis: unknown command " + cmd,
				})
			}
		}
		offset += len(line)
	}
	return issues
}

func validateMongo(text string) []Issue {
	methods := suggest.MongoMethodSet()
	trimmed := strings.TrimSpace(text)
	// Only validate the shell call shape db.<collection>.<method>(...).
	if !strings.HasPrefix(trimmed, "db.") {
		return nil
	}
	open := strings.IndexByte(trimmed, '(')
	if open < 0 {
		return nil // still being typed
	}
	head := trimmed[:open]
	parts := strings.Split(head, ".")
	if len(parts) >= 3 {
		method := parts[len(parts)-1]
		if method != "" && !methods[strings.ToUpper(method)] {
			off := strings.LastIndex(text, method)
			return []Issue{{Offset: max0(off), Length: len(method), Severity: SevWarning, Message: "mongo: unknown method " + method}}
		}
	}
	return nil
}

// unbalancedDelimiters reports the first unmatched (){}[] or unterminated string.
func unbalancedDelimiters(text string) []Issue {
	var stack []struct {
		ch  byte
		pos int
	}
	pairs := map[byte]byte{')': '(', ']': '[', '}': '{'}
	var quote byte
	quoteStart := 0
	for i := 0; i < len(text); i++ {
		c := text[i]
		if quote != 0 {
			if c == quote {
				quote = 0
			}
			continue
		}
		switch c {
		case '\'', '"', '`':
			quote = c
			quoteStart = i
		case '(', '[', '{':
			stack = append(stack, struct {
				ch  byte
				pos int
			}{c, i})
		case ')', ']', '}':
			if len(stack) == 0 || stack[len(stack)-1].ch != pairs[c] {
				return []Issue{{Offset: i, Length: 1, Severity: SevWarning, Message: "unbalanced " + string(c)}}
			}
			stack = stack[:len(stack)-1]
		}
	}
	if quote != 0 {
		return []Issue{{Offset: quoteStart, Length: 1, Severity: SevWarning, Message: "unterminated string"}}
	}
	if len(stack) > 0 {
		top := stack[len(stack)-1]
		return []Issue{{Offset: top.pos, Length: 1, Severity: SevWarning, Message: "unclosed " + string(top.ch)}}
	}
	return nil
}

func max0(i int) int {
	if i < 0 {
		return 0
	}
	return i
}

File: src/validate.rs
//! Non-blocking, never-false-positive query checks. Port of `internal/validate`.
//! Pairs a universal balanced-delimiter scan with the official Redis/Mongo
//! command sets; issues are advisory (shown as warnings, not blocking).

use crate::config::Driver;
use crate::suggest;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Severity {
    Warning,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Issue {
    pub offset: usize,
    pub length: usize,
    pub severity: Severity,
    pub message: String,
}

pub fn validate(driver: Driver, text: &str) -> Vec<Issue> {
    if text.trim().is_empty() {
        return Vec::new();
    }
    let mut issues = unbalanced_delimiters(text);
    match driver {
        Driver::Redis => issues.extend(validate_redis(text)),
        Driver::Mongo => issues.extend(validate_mongo(text)),
        _ => {}
    }
    issues
}

fn validate_redis(text: &str) -> Vec<Issue> {
    let commands = suggest::redis_command_set();
    let mut issues = Vec::new();
    let mut offset = 0usize;
    for line in text.split_inclusive('\n') {
        let trimmed = line.trim();
        if !trimmed.is_empty() && !trimmed.starts_with('#') {
            let first = trimmed.split_whitespace().next().unwrap_or("");
            let cmd = first.to_uppercase();
            if !commands.contains(&cmd) {
                let start = offset + line.find(first).unwrap_or(0);
                issues.push(Issue {
                    offset: start,
                    length: first.len(),
                    severity: Severity::Warning,
                    message: format!("redis: unknown command {cmd}"),
                });
            }
        }
        offset += line.len();
    }
    issues
}

fn validate_mongo(text: &str) -> Vec<Issue> {
    let methods = suggest::mongo_method_set();
    let trimmed = text.trim();
    if !trimmed.starts_with("db.") {
        return Vec::new();
    }
    let Some(open) = trimmed.find('(') else { return Vec::new() };
    let head = &trimmed[..open];
    let parts: Vec<&str> = head.split('.').collect();
    if parts.len() >= 3 {
        let method = *parts.last().unwrap();
        if !method.is_empty() && !methods.contains(&method.to_uppercase()) {
            let off = text.rfind(method).unwrap_or(0);
            return vec![Issue {
                offset: off,
                length: method.len(),
                severity: Severity::Warning,
                message: format!("mongo: unknown method {method}"),
            }];
        }
    }
    if let Err(err) = crate::db::mongo::parse_mongosh_command(trimmed) {
        return vec![Issue {
            offset: text.find(trimmed).unwrap_or(0),
            length: trimmed.len().max(1),
            severity: Severity::Error,
            message: format!("mongo: {err}"),
        }];
    }
    Vec::new()
}

fn unbalanced_delimiters(text: &str) -> Vec<Issue> {
    let bytes = text.as_bytes();
    let mut stack: Vec<(u8, usize)> = Vec::new();
    let mut quote = 0u8;
    let mut quote_start = 0usize;
    let opener = |c: u8| match c {
        b')' => b'(',
        b']' => b'[',
        b'}' => b'{',
        _ => 0,
    };
    for (i, &c) in bytes.iter().enumerate() {
        if quote != 0 {
            if c == quote {
                quote = 0;
            }
            continue;
        }
        match c {
            b'\'' | b'"' | b'`' => {
                quote = c;
                quote_start = i;
            }
            b'(' | b'[' | b'{' => stack.push((c, i)),
            b')' | b']' | b'}' => {
                if stack.last().map(|(ch, _)| *ch) != Some(opener(c)) {
                    return vec![Issue {
                        offset: i,
                        length: 1,
                        severity: Severity::Warning,
                        message: format!("unbalanced {}", c as char),
                    }];
                }
                stack.pop();
            }
            _ => {}
        }
    }
    if quote != 0 {
        return vec![Issue {
            offset: quote_start,
            length: 1,
            severity: Severity::Warning,
            message: "unterminated string".into(),
        }];
    }
    if let Some((ch, pos)) = stack.last() {
        return vec![Issue {
            offset: *pos,
            length: 1,
            severity: Severity::Warning,
            message: format!("unclosed {}", *ch as char),
        }];
    }
    Vec::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn balanced_ok() {
        assert!(validate(Driver::Mysql, "select * from t where (a=1)").is_empty());
    }

    #[test]
    fn detects_unbalanced() {
        let issues = validate(Driver::Mysql, "select * from t where (a=1");
        assert_eq!(issues.len(), 1);
        assert!(issues[0].message.contains("unclosed"));
    }

    #[test]
    fn detects_unterminated_string() {
        let issues = validate(Driver::Mysql, "select 'abc");
        assert_eq!(issues[0].message, "unterminated string");
    }

    #[test]
    fn redis_unknown_command_warns() {
        let issues = validate(Driver::Redis, "FOObar key");
        assert!(issues.iter().any(|i| i.message.contains("unknown command")));
    }

    #[test]
    fn redis_known_command_ok() {
        assert!(validate(Driver::Redis, "GET key").is_empty());
    }

    #[test]
    fn mongo_unknown_method_warns() {
        let issues = validate(Driver::Mongo, "db.users.frobnicate({})");
        assert!(issues.iter().any(|i| i.message.contains("unknown method")));
    }

    #[test]
    fn mongo_malformed_filter_is_error() {
        let issues = validate(Driver::Mongo, r#"db.users.find({profile.name: "Ada"})"#);

        assert_eq!(issues[0].severity, Severity::Error);
        assert!(issues[0].message.contains("profile.name"));
    }

    #[test]
    fn mongo_quoted_dotted_filter_is_ok() {
        let issues = validate(Driver::Mongo, r#"db.users.find({"profile.name": "Ada"})"#);

        assert!(issues.is_empty(), "{issues:?}");
    }
}
